### ISSUE TITLE
feat(auth): Integrate OAuth 2.0 for access to API

### ISSUE THREAD

**Issue by raisedadead** [freeCodeCamp/freeCodeCamp#16104]
We need to implement OAuth 2.0, with the below checklist:

- [ ] Investigate adding OAuth 2.0 and cost of implementation
  - [ ] Investigate schema and access control
  - [ ] Investigate creating / storing app identity

- [ ] Check integration via https://github.com/strongloop/loopback-component-oauth2

     **Update:** My intial investigation shows they are not maintaining it well and it seems to break with LoopBack@^3
       Refer https://github.com/strongloop/loopback-component-oauth2/issues/72

- [ ] Check integration via https://github.com/jaredhanson/oauth2orize (from passport author)
- [ ] ...

Note: Some ideas were previously discussed in https://github.com/freeCodeCamp/freeCodeCamp/issues/15914, this issue is tracking the implementation to prepare our code base for those ideas. Feel free to add ideas to original issue instead and keep this for implementation details only.

TL;DR: This mostly discusses the technical aspects of the implementation.

**Comment by BerkeleyTrue** [comment]
@raisedadead I'd like to see a separate of our oauth platform and core. This way they can move independently.

I'm not a big fan of relying on loopback anymore than we do. This whole loopback 2->3 business has really soured me on this platform. I'd rather we custom built then rely on unsupported platforms. This is also not as difficult, so I have been told.

**Comment by raisedadead** [comment]
@BerkeleyTrue Yup, that's totally awesome, and agreed. 

I have seen the frustration first hand while trying to upgrade, and can relate to your advice. Are you okay with some draft proposals then about a bit of rearchitecting (with backward support) the platform then?

I mean we do plan to have an independent static web-app and a api-only backend? Is that right?

**Comment by BerkeleyTrue** [comment]
@raisedadead yes the dream is a static site and a backend api. It won't be completely separate as that would make dev too complex to support new contributors. 

Also, there are a couple things I am investigating that would help us move away from loopback and are tangentially relevant to this issue. 

We will most likely move to a graph db (I'm currently talking to the Neo4j peeps). This will support our future as a platform for learning. We will be supporting relationships between campers and third party software as well as core.

We will add Graphql api, first internally to phase out the rest api and the custom endpoints we have written for the react app. Secondly as our public api for third party software to use. This will most likely be a separate platform that will authenticate with core.

Points on the main platform. I favor a functional style platform vs class based. I've had in mind something equivalent to Clojure's compojure/ring, but I'd hate to leave the well supported Expressjs.
